File: sqlite-android-studio-plugin/src/main/java/com.alecstrong.sqlite.android/psi/ColumnNameElementRef.java
package com.alecstrong.sqlite.android.psi;

import com.alecstrong.sqlite.android.SQLiteParser;
import com.alecstrong.sqlite.android.lang.SqliteFile;
import com.alecstrong.sqlite.android.lang.SqliteTokenTypes;
import com.alecstrong.sqlite.android.util.SqliteRenameUtil;
import com.intellij.psi.PsiElement;
import com.intellij.psi.tree.IElementType;
import com.intellij.psi.util.PsiTreeUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class ColumnNameElementRef extends SqliteElementRef {
  private TableNameElement leftTableDef;

  public ColumnNameElementRef(IdentifierElement idNode, String ruleName) {
    super(idNode, ruleName);
  }

  @Override public Class<? extends SqliteElement> identifierParentClass() {
    return ColumnNameElement.class;
  }

  @Override public IElementType identifierDefinitionRule() {
    return SqliteTokenTypes.RULE_ELEMENT_TYPES.get(SQLiteParser.RULE_column_def);
  }

  @NotNull @Override public Object[] getVariants() {
    leftTableDef = PsiTreeUtil.getPrevSiblingOfType(
        PsiTreeUtil.getParentOfType(getElement(), ColumnNameElement.class), TableNameElement.class);

    return super.getVariants();
  }

  @Nullable @Override public PsiElement resolve() {
    ColumnNameElement columnName =
        PsiTreeUtil.getParentOfType(getElement(), ColumnNameElement.class);
    if (columnName != null
        && columnName.getParent().getNode().getElementType() == identifierDefinitionRule()) {
      // If this is already a column definition return ourselves.
      return columnName;
    }
    leftTableDef = PsiTreeUtil.getPrevSiblingOfType(columnName, TableNameElement.class);

    return super.resolve();
  }

  @Override public PsiElement handleElementRename(final String newElementName) {
    SqliteFile file = (SqliteFile) myElement.getContainingFile();

    SqliteRenameUtil.SqliteUsageInfo usageInfo =
        SqliteRenameUtil.findUsages(myElement, newElementName, file);
    SqliteRenameUtil.doRename(myElement, newElementName, usageInfo, file, null);

    return myElement;
  }

  @Override public boolean isAccepted(PsiElement element) {
    if (leftTableDef == null) {
      // There was no table specified, all table names and column names are valid.
      return super.isAccepted(element) || (element instanceof TableNameElement
          && element.getParent().getNode().getElementType()
          == SqliteTokenTypes.RULE_ELEMENT_TYPES.get(SQLiteParser.RULE_create_table_stmt));
    } else {
      return super.isAccepted(element) && leftTableDef.isSameTable(
          PsiTreeUtil.getChildOfType(element.getParent().getParent(), TableNameElement.class));
    }
  }
}

File: sqlite-android-studio-plugin/src/main/kotlin/com/squareup/sqlite/android/psi/ColumnNameElementRef.kt
package com.squareup.sqlite.android.psi

import com.intellij.psi.PsiElement
import com.squareup.sqlite.android.SQLiteParser
import com.squareup.sqlite.android.lang.SqliteFile
import com.squareup.sqlite.android.lang.SqliteTokenTypes.RULE_ELEMENT_TYPES
import com.squareup.sqlite.android.psi.SqliteElement.ColumnNameElement
import com.squareup.sqlite.android.psi.SqliteElement.TableNameElement
import com.squareup.sqlite.android.util.childOfType
import com.squareup.sqlite.android.util.doRename
import com.squareup.sqlite.android.util.elementType
import com.squareup.sqlite.android.util.findUsages
import com.squareup.sqlite.android.util.parentOfType
import com.squareup.sqlite.android.util.prevSiblingOfType

internal class ColumnNameElementRef(idNode: IdentifierElement, ruleName: String)
: SqliteElementRef(idNode, ruleName) {
  private var leftTableDef: TableNameElement? = null

  override protected val identifierDefinitionRule = RULE_ELEMENT_TYPES[SQLiteParser.RULE_column_def]

  override fun getVariants(): Array<Any> {
    setLeftTable()
    return super.getVariants()
  }

  override fun resolve(): PsiElement? {
    val columnName = element.parentOfType<ColumnNameElement>()
    if (columnName != null && columnName.parent.elementType === identifierDefinitionRule) {
      // If this is already a column definition return ourselves.
      return columnName
    }
    setLeftTable()

    return super.resolve()
  }

  override fun handleElementRename(newElementName: String): PsiElement {
    val file = myElement.containingFile as SqliteFile

    val usageInfo = myElement.findUsages(newElementName)
    myElement.doRename(newElementName, usageInfo, file, null)

    return myElement
  }

  override fun isAccepted(element: PsiElement) =
      when (leftTableDef) {
        null -> super.isAccepted(element) || element is TableNameElement
            && element.getParent().elementType === RULE_ELEMENT_TYPES[SQLiteParser.RULE_create_table_stmt]
        else -> super.isAccepted(element)
            && leftTableDef!!.isSameTable(element.parent.parent.childOfType<TableNameElement>())
      }

  private fun setLeftTable() {
    leftTableDef = element.parentOfType<ColumnNameElement>()?.prevSiblingOfType<TableNameElement>()
  }
}
